File: src/http-client.ts
import { AstJson } from "@sdkgen/parser";
import { randomBytes } from "crypto";
import { request as httpRequest } from "http";
import { request as httpsRequest } from "https";
import { hostname } from "os";
import { Context } from "./context";
import { decode, encode } from "./encode-decode";

interface ErrClasses {
    [className: string]: any
}

export class SdkgenHttpClient {
    private baseUrl: URL
    extra = new Map<string, any>();

    constructor(baseUrl: string, private astJson: AstJson, private errClasses: ErrClasses) {
        this.baseUrl = new URL(baseUrl);
    }

    async makeRequest(ctx: Context | null, functionName: string, args: any) {
        const func = this.astJson.functionTable[functionName];
        if (!func) {
            throw new Error(`Unknown function ${functionName}`);
        }

        const request = {
            version: 3,
            requestId: ctx && ctx.request? ctx.request.id + randomBytes(6).toString("hex") : randomBytes(16).toString("hex"),
            name: functionName,
            args: encode(this.astJson.typeTable, `${functionName}.args`, func.args, args),
            extra: {
                ...this.extra,
                ...(ctx && ctx.request ? ctx.request.extra : {})
            },
            deviceInfo: ctx && ctx.request? ctx.request.deviceInfo : {
                id: hostname(),
                type: "node"
            },
        };

        const options = {
            hostname: this.baseUrl.hostname,
            path: this.baseUrl.pathname,
            port: this.baseUrl.port,
            method: "POST",
        };

        const encodedRet = await new Promise<any>((resolve, reject) => {
            const req = (this.baseUrl.protocol === "http:" ? httpRequest : httpsRequest)(options, res => {
                let data = "";
                res.on("data", chunk => {
                    data += chunk;
                });
                res.on("end", () => {
                    try {
                        const response = JSON.parse(data);

                        if (response.error) {
                            reject(response.error);
                        } else {
                            resolve(response.result);
                        }
                    } catch (e) {
                        reject({ type: "Fatal", message: e.toString() });
                    }
                });

            });

            req.on("error", (e) => {
                console.error(`problem with request: ${e.message}`);
                reject({ type: "Fatal", message: e.toString() });
            });

            req.write(JSON.stringify(request));
            req.end();
        }).catch(err => {
            const errClass = this.errClasses[err.type];
            if (errClass)
                throw new errClass(err.message);
            else
                throw err;
        });

        return decode(this.astJson.typeTable, `${functionName}.ret`, func.ret, encodedRet);
    }
}

File: lib/http_client.dart
import 'package:http/http.dart' as http;
import 'package:convert/convert.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';
import 'dart:math';

import 'types.dart';

class SdkgenError implements Exception {
  String message;
  SdkgenError(this.message);
}

class SdkgenHttpClient {
  String baseUrl;
  Map<String, Object> typeTable;
  Map<String, FunctionDescription> fnTable;
  Map<String, Function> errTable;
  String deviceId;
  Random random = Random.secure();

  SdkgenHttpClient(this.baseUrl, this.typeTable, this.fnTable, this.errTable);

  _randomBytesHex(int bytes) {
    return hex.encode(List<int>.generate(bytes, (i) => random.nextInt(256)));
  }

  _throwError(String type, String message) {
    var factory = errTable[type] == null ? errTable["Fatal"] : errTable[type];
    throw Function.apply(factory, [message]);
  }

  _deviceId() async {
    if (deviceId == null) {
      SharedPreferences prefs = await SharedPreferences.getInstance();
      if (prefs.containsKey("sdkgen_deviceId")) {
        deviceId = prefs.getString("sdkgen_deviceId");
      } else {
        deviceId = _randomBytesHex(16);
        prefs.setString("sdkgen_deviceId", deviceId);
      }
    }
    return deviceId;
  }

  Future<Object> makeRequest(
      String functionName, Map<String, Object> args) async {
    try {
      var func = fnTable[functionName];
      var encodedArgs = Map();
      args.forEach((argName, argValue) {
        encodedArgs[argName] = encode(typeTable, "$functionName.args.$argName",
            func.args[argName], argValue);
      });

      var body = {
        "version": 3,
        "requestId": _randomBytesHex(16),
        "name": functionName,
        "args": encodedArgs,
        "extra": {},
        "deviceInfo": {
          "id": await _deviceId(),
          "type": "dart"
        }
      };

      var response = await http.post(baseUrl, body: jsonEncode(body));
      var responseBody = jsonDecode(utf8.decode(response.bodyBytes));

      if (responseBody["error"] != null) {
        throw _throwError(
            responseBody["error"]["type"], responseBody["error"]["message"]);
      } else {
        return decode(
            typeTable, "$functionName.ret", func.ret, responseBody["result"]);
      }
    } catch (e) {
      if (e is SdkgenError)
        throw e;
      else
        throw _throwError("Fatal", e.toString());
    }
  }
}
